src/vpn-api/worker.ts

```diff
@@ -71,91 +71,245 @@ const getInstanceSubnet = (instanceId: number) => {
+							),
+						);
+				} else {
+					logger.debug(
+						`authentication failed for client_id=${clientId} uuid=${
+							data.username
+						}`,
+					);
+					metrics.inc(Metrics.AuthFailures);
+					return vpn
+						.exec(`client-deny ${clientId} ${keyId} "AUTH_FAILURE"`)
+						.then(() =>
+							logger.info(
+								`rejected client_id=${clientId} uuid=${
+									data.username
+								} reason=AUTH_FAILURE`,
+							),
+						);
+				}
+			})
+			.catch(err => {
+				logger.alert(`api error: ${err}\n${err.stack}`);
+				vpn
+					.exec(`client-deny ${clientId} ${keyId} "AUTH_ERROR"`)
+					.then(() =>
+						logger.info(
+							`rejected client_id=${clientId} uuid=${
+								data.username
+							} reason=AUTH_ERROR`,
+						),
+					);
+			});
+	});
+
+	vpn.on('client:established', (clientId, data) => {
+		logger.info(
+			`connection established with client_id=${clientId} uuid=${data.username}`,
+		);
+		metrics.inc(Metrics.OnlineDevices);
+		metrics.inc(Metrics.TotalDevices);
+		clientLog[clientId] = Date.now();
+		return clients.connected(data).then(logStateUpdate);
 	});
 
-	const app = (Promise.promisifyAll(express()) as any) as AsyncApplication;
-	app.disable('x-powered-by');
-	app.get('/ping', (_req, res) => res.send('OK'));
-	app.use(morgan('combined'));
-	app.use(compression());
-	app.use(apiFactory());
-	app.use(Raven.errorHandler());
-
-	return app
-		.listenAsync(apiPort)
-		.tap(() =>
-			logger.info(
-				`open-balena-vpn worker-${instanceId} listening on port ${apiPort}`,
-			),
-		)
-		.tap(() => openvpn.start())
-		.tap(() =>
-			net.createConnection('/var/run/haproxy.sock', function(this: net.Socket) {
-				this.on('error', err => {
-					logger.error(`Error connecting to haproxy socket: ${err.message}`);
-					process.exit(1);
-				});
-				const preamble = `set server vpn-cluster/vpn${instanceId}`;
-				this.write(
-					`${preamble} addr 127.0.0.1 port ${vpnPort}\r\n${preamble} state ready\r\n`,
-					() => this.destroy(),
-				);
-			}),
-		)
-		.return(true);
+	vpn.on('client:bytecount', (clientId, data) => {
+		writeBitrateMetrics(data, Date.now() - clientLog[clientId]);
+	});
+
+	vpn.on('client:disconnect', (clientId, data) => {
+		let msg = `session ended for`;
+		if (data.username == null) {
+			return logger.debug(`${msg} rejected client_id=${clientId}`);
+		}
+		msg = `${msg} client_id=${clientId} uuid=${data.username}`;
+		if (
+			data.bytes_received != null &&
+			data.bytes_sent != null &&
+			data.time_duration > 0
+		) {
+			msg = `${msg} (len=${data.time_duration} rx=${data.bytes_received} tx=${
+				data.bytes_sent
+			})`;
+			metrics.histogram(Metrics.SessionRxBytes, data.bytes_received);
+			metrics.histogram(Metrics.SessionTxBytes, data.bytes_sent);
+			writeBitrateMetrics(data as VpnClientBytecountData, data.time_duration);
+		}
+		logger.info(msg);
+		metrics.dec(Metrics.OnlineDevices);
+		delete clientLog[clientId];
+		return clients.disconnected(data).then(logStateUpdate);
```

Comment:
I don't think it can result in a rejection… If you take a look at the code in `clients.ts` it has a catch on itself which calls itself after a delay.

As you noted later in the review, I'm now returning the final promise from the client.{connect, disconnect} functions. Prior to this PR, the returned promise from these calls was being ignored anyway.